### PR TITLE
[FIX] website_sale(_loyalty)_delivery: set correct delivery prices

### PR DESCRIPTION
**Pull request by kawkb**
Issue:

- when we have a shipping method that has a discount the shipping and the delivery prices are not displayed correctly (see attached screenshot).

Steps To Reproduce:

- configure the sale app to have the discount&loyalty and free shipping options enabled.
- create a product that cost 100$.
- create a new discount/loyalty where type is promotions and set rule to if minimum 300$ spent the reward is free shipping max to 75$.
- create a shipping method based on rules where the delivery product is a service that costs 100$.
- in your shipping mehtod pricing add a line with if quantity >= 3.0 then fixed price 100.0 and publish it.
- in your website shop select 3 items from the product you created. and select your delivery method.
- notice the prices not updating correctly.

Solution:
- Before the fix we always assumed that free_shipping_lines are
equal to 0.

- free_shipping_lines might be shipping discounts not actual free
shipping.so I adjusted update_eshop_carrier method to make it handle the
 case of discounted shipping.

opw-3718885

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by kawkb** [comment]
Both the shipping method price tag and the delivery price are 0 (should be 100 and 25)

image

**Review by nle-odoo — APPROVED** [review]
looks good to me

**Review by vava-odoo — CHANGES_REQUESTED** [review]
Hi @kawkb,

Thanks for the fix and the test.
I'm not familiar with the combo with loyalty - delivery - ecommerce, but the fix seems to work fine.

I personnally find it weird that the badge value differs from what is in the sum-up on the right. I'd say the badge could also show $25 in your example. @paju-odoo could you confirm?

Cheers!
(and good luck with the forward-port :eyes:)

**Comment by paju-odoo** [comment]
Hi @kawkb,

As discussed previously, the badge and the delivery in the sum-up are displayed correctly (100 and 25 respectively). However, the shipping discount in the product summary was supposed to display the discount (-75 in this case): https://i.imgur.com/M5DzUzh.png + https://i.imgur.com/KcuAnUA.png. Could you fix that? If not, it's indeed a bit odd to have different values on the badge and in the sum-up. 
While testing, I also encountered another issue: the free shipping product appears twice in the product summary. Any idea of the reason? I followed the steps explained in this PR to do the configuration and obtained this result
Video of the flow: https://drive.google.com/file/d/12aI7MHiA6H1iSQUgjcYk-xMuDHhN3wMh/view?usp=sharing

cc @vava-odoo

**Comment by kawkb** [comment]
Hello @paju-odoo

Thank you for your review, As discussed previously, this PR fix the 1 and 3 fields. the 2 field might be best suited for someone from the R&D team due to its complexity and my unfamiliarity with certain parts of the code.

I remember @chevalierv mentioning that the team would provide suggestions during the PR review, and I look forward to any guidance or modifications they suggest.

File: addons/website_sale_loyalty_delivery/tests/test_website_sale_loyalty_delivery.py [hunk 1]
```diff
@@ -99,3 +99,45 @@ def test_shop_sale_gift_card_keep_delivery(self):
...
+            'website_published': True,
+            'product_id': product_paid_delivery.id,
+        })
+        self.env['delivery.price.rule'].create([{
+            'carrier_id': delivery_with_rule.id,
+            'variable': 'quantity',
+            'operator': '>=',
+            'max_value': 3,
+            'list_base_price': 100,
+        }])
```
[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/website_sale_loyalty_delivery/tests/test_website_sale_loyalty_delivery.py]: @vava-odoo Anv is sad that you still use magic numbers :(

**Review by chevalierv — COMMENTED** [review]
Hello @kawkb 

Ultimately, I've pushed some changes directly in your PR because the suggestion was too far outside the diff in this PR. 

I've left a few comments to explain why I've done what.

Let me know if it's ok for you.

Have a nice day

**Comment by chevalierv** [comment]
@robodoo r+

File: addons/website_sale_loyalty_delivery/tests/test_website_sale_loyalty_delivery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vava-odoo) [reply addons/website_sale_loyalty_delivery/tests/test_website_sale_loyalty_delivery.py]: And I'm sad you don't provide reasons to the security team to accept your changes

**Comment by AntoineVDV** [comment]
@robodoo override=ci/security
Reason: `innerHtml`: same logic as before (computed value unaltered by the user), but lines have been grouped.
Link to ci/security build (16.0): https://runbot.odoo.com/runbot/build/60716178

**Comment by chevalierv** [comment]
@robodoo r+

**Pull request closed without merging**